### ISSUE TITLE
How to properly scale with multiple event conditions

### ISSUE THREAD

**Issue by LuggiStruggi** [patrick-kidger/diffrax#637]
Hi diffrax team,

I am working on a problem, where the amount of cond_fn scales with my state dimension. Unfortunately, this currently scales pretty badly timewise even when jitting the code. Do you have any idea what could be done to improve here?

Here as an example:

```math
\dot{y} + y = 0
```
```math
y(0) \sim [0.9, 1.0]
```
event condition (any state):
```math
y - 0.1 = 0
```
 Solving with Euler and a fixed stepsize of 0.005ms between 0 and 30ms (or until first event), it scales really badly with the state dimension:

```
   Dim |   Time (s)
--------------------
     1 |   0.001022
     2 |   0.000625
     4 |   0.000657
     8 |   0.001162
    16 |   0.001870
    32 |   0.004110
    64 |   0.010566
   128 |   0.044464
   200 |   0.218422
```

see for comparison, the timing where i just used the cond_fn on the first element of the state `y[0]`:

```
   Dim |   Time (s)
--------------------
     1 |   0.000723
     2 |   0.000612
     4 |   0.000595
     8 |   0.000615
    16 |   0.000596
    32 |   0.000716
    64 |   0.000903
   128 |   0.000724
   200 |   0.001835
```

Do you see any way to speed this up a reasonable amount?

Thanks for the help!

Here is the MWE i used (Single cond_fn commented out)

```python
import jax
import jax.numpy as jnp
import diffrax as dfx
import optimistix as optx
from timeit import default_timer as timer

def solve_once(y0, t0, t1, dt0, solver, stepsize_controller, event):
    term = dfx.ODETerm(lambda t, y, args: -y)
    return dfx.diffeqsolve(
        terms=term,
        solver=solver,
        t0=t0,
        t1=t1,
        dt0=dt0,
        y0=y0,
        args=None,
        stepsize_controller=stepsize_controller,
        event=event,
        saveat=dfx.SaveAt(t0=False, t1=True, steps=False),
        max_steps=3000000,
        throw=True,
    )

solve_once = jax.jit(solve_once, static_argnames=["solver", "stepsize_controller", "event"])

def benchmark_diffeqsolve_with_event():
    key = jax.random.PRNGKey(0)
    t0, t1, dt0 = 0.0, 30.0, 0.005
    solver = dfx.Euler()
    controller = dfx.ConstantStepSize()
    dims = [1, 2, 4, 8, 16, 32, 64, 128, 200]

    print(f"{'Dim':>6} | {'Time (s)':>10}")
    print("-" * 20)

    for n in dims:
        key, subkey = jax.random.split(key)
        y0 = jax.random.uniform(subkey, shape=(n,), minval=0.9, maxval=1.1)

        cond_fns = [lambda t, y, *args, i=i, **kwargs: y[i] - 0.01 for i in range(n)]
        # cond_fns = lambda t, y, *args, **kwargs: y[0] - 0.01

        event = dfx.Event(cond_fns, root_finder=optx.Newton(rtol=1e-4, atol=1e-4))

        _ = solve_once(y0, t0, t1, dt0, solver, controller, event)  # warm-up

        start = timer()
        _ = solve_once(y0, t0, t1, dt0, solver, controller, event)
        end = timer()

        print(f"{n:6d} | {end - start:10.6f}")

benchmark_diffeqsolve_with_event()

```

**Comment by patrick-kidger** [comment]
I think scaling with the number of condition functions is probably inevitable unfortunately.

But a single vectorized condition function should already be totally doable as a user! Untested but something like this should probably work for you:
```python
def cond_fn(t, y, args, **kwargs):
    return jnp.min((y - 0.1)**2)
```
In this case you just need a function that moves smoothly to 0 when any of your components are near 0.1.

**Comment by LuggiStruggi** [comment]
Hi Patrick,

```python
def cond_fn(t, y, args, **kwargs):
    return jnp.prod(y - 0.1)
```
already seems to speed things up alot. One big issue here is this way I won't be able to tell which element of `y` caused the event. Another problem is that with an increase in the y dimension and multiple ys below 1, i get an event at the wrong time as for a sufficiently big N and lets say all ys being 0.15
```math
(0.15 - 0.1)^N \approx  0
```
also
```
def cond_fn(t, y, args, **kwargs):
    return jnp.min((y - 0.1)**2)
```
will never work, as it never crosses 0, which I think needs to happen for an Event to be caused (please correct me if I'm wrong. It just seems like it from the results I am getting). Anyways
```python
def cond_fn(t, y, args, **kwargs):
    return jnp.min(y - 0.1)
```
seems to work, but back to the issue that i won't know which element caused the spike.

I could do smth like
```python
event_idx = jnp.argmin(sol.ys - 0.1)
```
afterwards but that seems a bit duplicate. It would be nice if the event mask we are getting back could also have elements that are vectors if the respective cond_fn was vector valued as well. Or do you think that is unnecessary? It would be certainly more elegant to define for some problems than doing a list comprehension of lambdas or smth similar over all y dimensions.

Do you think that this extension is very difficult to implement?
Thanks for your Help! :)

**Comment by LuggiStruggi** [comment]
Also I feel like there could be a potential problem in the current code (I might be wrong) which happens, if we have two cond_fns which cross 0 at the same excact ODE-solver step, but the second one in the passed cond_fn PyTree is the one that in reality happens first. In that case the actual (in real time) first cond_fn ist just ignored. What would need to be done to be exact is to run a root solver for each cond_fn that crossed 0 and then check which root appeared first and return the state at that time as the solution.

**Comment by LuggiStruggi** [comment]
Okay! The only worry I have with this approach is smh similar to https://github.com/patrick-kidger/diffrax/issues/639. In case of multiple roots, how can i be certain the root finder will pick up the very first? Any idea on how to handle that? :) Thanks for the help!

**Comment by LuggiStruggi** [comment]
Thanks! Yes that seems really reasonable.

Edit: I'm struggling to find a function that has both: this unique first root + a sign change at 0 crossing
